src/digraph.rs

```diff
@@ -2761,3 +2774,50 @@ impl PyGCProtocol for PyDiGraph {
         self.node_removed = false;
     }
 }
+
+fn _from_adjacency_matrix<'p, T>(
+    py: Python<'p>,
+    matrix: PyReadonlyArray2<'p, T>,
+    null_value: T,
+) -> PyDiGraph
+where
+    T: Copy + std::cmp::PartialEq + numpy::Element + pyo3::ToPyObject + IsNan,
+{
+    let array = matrix.as_array();
+    let shape = array.shape();
+    let mut out_graph = StableDiGraph::<PyObject, PyObject>::new();
+    let _node_indices: Vec<NodeIndex> = (0..shape[0])
+        .map(|node| out_graph.add_node(node.to_object(py)))
+        .collect();
+    array
+        .axis_iter(Axis(0))
+        .enumerate()
+        .for_each(|(index, row)| {
+            let source_index = NodeIndex::new(index);
+            for target_index in 0..row.len() {
```

Comment:
One last comment before we merge this. It's not a blocker but while we are here, it might be better to iterate over the elements in `row` like:

```suggestion
            for (target_index, elem) in row.iter().enumerate() {
```
to avoid indexing `row[[target_index]]`.

Reply:
Good idea, done in: https://github.com/Qiskit/retworkx/pull/412/commits/ada39eaa25fc5583e9506548e4495fc52e46bfb2